Fixes #4464 - policy for foreman-proxy

lzap opened the pull request:
This is preliminary policy for proxy, feel free to review and _test_.
Instructions how to easily test will follow.

What was tested so far:
- create host
- edit host
- rebuild host
- delete host
- list puppetca certificates

PLEASE I NEED YOUR ASSISTANCE WITH HELPING TO TEST THIS POLICY. It is piece of
cake to test it, won't break your existing instance at all. Read the
instructions in my comments bellow.

lzap committed:
Fixes #4464 - policy for foreman-proxy

lzap commented:
Instructions how to test on RHEL6 (RHEL7 is not supported yet):

```
git clone https://github.com/lzap/foreman-selinux -b proxy-4464
cd foreman-selinux
rm -f foreman.{fc,if,te}
sed -i s/@@VERSION@@/99.9/ *.te
make -f /usr/share/selinux/devel/Makefile load DISTRO=rhel6
```

Restart foreman-proxy and do your testing. Check for denials, leave denials in the comments here:

```
ausearch -m AVC -m USER_AVC
```

When done, delete the git repository, remove the SELinux module and restart foreman-proxy to get back to the normal state.

```
semodule -r foreman-proxy
```

Thank you!

lzap commented:
I pushed changes for foreman-proxy separate subpackage. Please review.

lzap commented:
Pushed last bunch of changes, working on SPEC changes.

domcleal commented on foreman.fc:
```diff
@@ -48,7 +48,7 @@
 # Foreman Tasks plugin
 
-/usr/share/gems/gems/foreman-tasks-*/bin/foreman-tasks -- gen_context(system_u:object_r:foreman_tasks_exec_t,s0)
+/usr/share/gems/gems/foreman-tasks-.*/bin/foreman-tasks -- gen_context(system_u:object_r:foreman_tasks_exec_t,s0)
```
Looks unrelated, probably useful in its own right.

lzap replied:
Yeah I am just correcting. Do you insist on splitting this one?

domcleal commented on foreman-proxy-selinux-relabel:
```diff
@@ -0,0 +1,11 @@
+  /etc/foreman-proxy \
+  /usr/share/foreman-proxy \
+  /var/lib/foreman-proxy \
+  /var/log/foreman-proxy \
```
Missing a file context or policy for /var/log/foreman-proxy, I get this on startup:

<pre>
type=AVC msg=audit(1413880810.218:1653): avc:  denied  { open } for  pid=32718 comm="ruby" path="/var/log/foreman-proxy/proxy.log" dev="vda1" ino=805962 scontext=system_u:system_r:foreman_proxy_t:s0 tcontext=system_u:object_r:var_log_t:s0 tclass=file
</pre>

lzap replied:
Oh I was testing with foreground mode. Fixing.

domcleal commented on common/selinux-relabel.pod.in.sh:
```diff
@@ -0,0 +1,24 @@
+=head1 DESCRIPTION
+
+The B<$PROG-selinux-relabel> program will call restorecon(8) on all
+$NAME files efectively restoring SELinux context.
```
effectively

domcleal commented on common/selinux-relabel.pod.in.sh:
```diff
@@ -0,0 +1,24 @@
+The B<$PROG-selinux-relabel> program will call restorecon(8) on all
+$NAME files efectively restoring SELinux context.
+
+The program is executed during installation of $NAME package
```
"during the installation of the $NAME SELinux package"

domcleal commented on common/selinux-enable.pod.in.sh:
```diff
@@ -0,0 +1,25 @@
+SELinux module. It defines some SELinux booeals which are required
+for operation.
+
+The program is executed during installation of $NAME package
```
ditto from relabel

domcleal commented on common/selinux-enable.pod.in.sh:
```diff
@@ -0,0 +1,25 @@
+=head1 DESCRIPTION
+
+The B<$PROG-selinux-enable> program will load $NAME
+SELinux module. It defines some SELinux booeals which are required
```
booleans

domcleal commented on common/selinux-disable.pod.in.sh:
```diff
@@ -0,0 +1,24 @@
+The B<$PROG-selinux-disable> program will unload and disable $NAME
+SELinux module. It also removes all the relevant user defined ports.
+
+The program is executed during installation of $NAME package
```
ditto relabel

domcleal commented on common/selinux-enable.pod.in.sh:
```diff
@@ -0,0 +1,25 @@
+
+=head1 SEE ALSO
+
+$PROG-selinux-relabel(8), $PROG-selinux-enable(8)
```
s/enable/disable/

lzap replied:
Fixed all the doco remarks, thanks.

domcleal commented on foreman-proxy.te:
```diff
@@ -0,0 +1,158 @@
+hostname_exec(foreman_proxy_t)
+puppet_read_lib(foreman_proxy_t)
+miscfiles_manage_cert_dirs(foreman_proxy_t)
+miscfiles_manage_cert_files(foreman_proxy_t)
```
is "manage" required?

lzap replied:
Can't remember (surpise! :-)

Let me switch that back to read and re-test everything.

domcleal commented on foreman-proxy.te:
```diff
@@ -0,0 +1,158 @@
+')
+files_manage_generic_tmp_files(foreman_proxy_t)
+
+# DNS PLUGIN
```
specifically the nsupdate provider?  What about virsh, or nsupdate_gss?

domcleal commented on foreman-proxy.te:
```diff
@@ -0,0 +1,158 @@
+# ping
+netutils_domtrans_ping(foreman_proxy_t)
+
+# DHCP PLUGIN
```
specifically isc?

lzap replied:
Re providers - indeed, this is the first cut. I am going to enclose everything into boolean blocks now. Will create some kind of `_generic` bools and then `_provider` bools. Will shortly send an update, I will create TODO comments for those providers without any support

domcleal commented on foreman-proxy.te:
```diff
@@ -0,0 +1,158 @@
+# nsupdate
+allow foreman_proxy_t self:process signal;
+# ping
+netutils_domtrans_ping(foreman_proxy_t)
```
part of DHCP I think, not DNS

domcleal commented:
1. I've got to admit, I'm nervous about this, I don't think we have a good idea of the possible behaviours of the proxy because it has so many configuration variables, but I suppose it has to be done at some point.
2. The policy should have everything behind booleans so you can lock down the behaviour depending on which modules are enabled.  We could later do this from the installer.

lzap commented:
1. I have no comment for this :-) I spent some time already and yeah, my testing did work well. That's the reason why I bring this PR over and over again - we should start testing to fit the 1.7 timeframe! I'd like to have this in nightlies ASAP.
2. Totally agree, you will see me adding these booleans after I do your comments.

Thanks for working on this, now I am going to work on your comments. @eLobato you are also testing this I've heard?

domcleal commented:
I think the best way to do this would be to merge it soon once the basics are fixed, but not make it the default until post-1.7, so if people want, they can opt-in and even supplement or fix the policy.

We can then create a test plan on the wiki (list all modules and providers, then tick them off) and once we're sure we've got decent coverage, make it installed by default via the installer or packages.

lzap commented:
Sounds like a good plan, let's not execute foreman-proxy-selinux-enable in RPMs and let know the users how to enable it manually. I am working on the policy today, I will be doing some bigger changes in the _admin interfaces. I hope to finish this week.

lzap commented:
Ok all your remarks done, policy was greatly rewritten (the optional bits) because it turned out that when I was returning to this long-standing PR, I did not execute foreman-proxy-selinux-relabel therefore I was testing against proxy running unconfined :-)

What was tested on both CentOS 6.6 and CentOS 7.0:
- ISC DHCP
- ISC BIND
- ISC TFTP

Fixed may errors, incorporated Doms comments. Works fine on RHEL6, testing RHEL7 now. What was covered so far: Create/remove host with ISC services (baremetal and libvirt).

That UDP thing (http://projects.theforeman.org/issues/8030) was taken here https://github.com/theforeman/foreman-selinux/pull/39.

lzap commented:
I am done with testing. Works on both CentOS 6 and 7.

Here is my suggestion: Let's merge this, I will prepare packages what will just install the policy, but won't turn it on. And then let's call for more help with testing.

lzap commented:
I've copied puppetca policy for both RHEL6 and RHEL7 until wrappers are delivered: https://github.com/theforeman/smart-proxy/pull/201

Also created one additional boolean for optional `sudo` support. Tested, please review.

domcleal commented on foreman-proxy.te:
```diff
@@ -0,0 +1,245 @@
+    ')
+')
+
+# PUPPET CA SUPPORT
```
should be tunable, it's a lot of permissions otherwise

lzap replied:
> should be tunable, it's a lot of permissions otherwise

Yeah missed, fixing.

## 

Later,
 Lukas #lzap Zapletal

domcleal commented on foreman-proxy.te:
```diff
@@ -0,0 +1,245 @@
+
+# server networking
+allow foreman_proxy_t self:tcp_socket { listen accept };
+corenet_tcp_bind_http_port(foreman_proxy_t)
```
websm_port_t too for katello's reconfiguration?

lzap replied:
> websm_port_t too for katello's reconfiguration?

Good catch!

## 

Later,
 Lukas #lzap Zapletal

domcleal commented on foreman-proxy.te:
```diff
@@ -0,0 +1,245 @@
+# </copy of puppetca_t policy>
+
+optional_policy(`
+    hostname_exec(foreman_proxy_t)
```
any idea what that's for?  My guess is a) running Puppet binaries, b) the facter plugin

lzap replied:
> any idea what that's for?  My guess is a) running Puppet binaries, b) the facter plugin

I really need to start documenting every single line, commenting out and
will find why this is needed.

## 

Later,
 Lukas #lzap Zapletal

domcleal commented on foreman-proxy.te:
```diff
@@ -0,0 +1,245 @@
+')
+optional_policy(`
+    corecmd_search_bin(foreman_proxy_t)
+    allow foreman_proxy_t sendmail_exec_t:file { getattr_file_perms execute };
```
`mta_sendmail_exec(foreman_proxy_t)` already?

lzap replied:
> `mta_sendmail_exec(foreman_proxy_t)` already?

Yeah, missed one interface. This line becomes

```
mta_sendmail_access_check(foreman_proxy_t)
```

and moved the bin_t search to ruby runtime block.

## 

Later,
 Lukas #lzap Zapletal

domcleal commented on foreman-proxy.te:
```diff
@@ -0,0 +1,245 @@
+# DHCP PLUGIN
+tunable_policy(`foreman_proxy_manage_dhcp_generic', `
+    # ping support via ECHO TCP service
+    ifdef(`distro_rhel7', `
```
and Fedora?

lzap replied:
> > -    ifdef(`distro_rhel7',`
> 
> and Fedora?

Yeah, I need to make sure that I only have special statements for RHEL6,
RHEL7&Fedora is more modern and they both fit in the else block.
Corrected with `ifndef`.

## 

Later,
 Lukas #lzap Zapletal

domcleal commented:
Seems to be missing support for the Puppet endpoint for class imports.

<pre>
time->Wed Nov 12 11:32:46 2014
type=SYSCALL msg=audit(1415791966.058:1062): arch=c000003e syscall=4 success=no exit=-13 a0=58d2ad0 a1=7f0f233cb7b0 a2=7f0f233cb7b0 a3=7f0f30b02c60 items=0 ppid=1 pid=23066 auid=4294967295 uid=997 gid=995 euid=997 suid=997 fsuid=997 egid=995 sgid=995 fsgid=995 tty=(none) ses=4294967295 comm="ruby" exe="/usr/bin/ruby" subj=system_u:system_r:foreman_proxy_t:s0 key=(null)
type=AVC msg=audit(1415791966.058:1062): avc:  denied  { search } for  pid=23066 comm="ruby" name="puppet" dev="vda1" ino=25173550 scontext=system_u:system_r:foreman_proxy_t:s0 tcontext=system_u:object_r:puppet_etc_t:s0 tclass=dir
----
time->Wed Nov 12 11:43:07 2014
type=SYSCALL msg=audit(1415792587.194:1064): arch=c000003e syscall=4 success=yes exit=0 a0=5908130 a1=7f0f233cb970 a2=7f0f233cb970 a3=7f0f30b02c60 items=0 ppid=1 pid=23344 auid=4294967295 uid=997 gid=995 euid=997 suid=997 fsuid=997 egid=995 sgid=995 fsgid=995 tty=(none) ses=4294967295 comm="ruby" exe="/usr/bin/ruby" subj=system_u:system_r:foreman_proxy_t:s0 key=(null)
type=AVC msg=audit(1415792587.194:1064): avc:  denied  { getattr } for  pid=23344 comm="ruby" path="/etc/puppet/puppet.conf" dev="vda1" ino=26647789 scontext=system_u:system_r:foreman_proxy_t:s0 tcontext=system_u:object_r:puppet_etc_t:s0 tclass=file
----
time->Wed Nov 12 11:43:07 2014
type=SYSCALL msg=audit(1415792587.199:1065): arch=c000003e syscall=2 success=yes exit=12 a0=59608a8 a1=80000 a2=1b6 a3=1 items=0 ppid=1 pid=23344 auid=4294967295 uid=997 gid=995 euid=997 suid=997 fsuid=997 egid=995 sgid=995 fsgid=995 tty=(none) ses=4294967295 comm="ruby" exe="/usr/bin/ruby" subj=system_u:system_r:foreman_proxy_t:s0 key=(null)
type=AVC msg=audit(1415792587.199:1065): avc:  denied  { open } for  pid=23344 comm="ruby" path="/etc/puppet/puppet.conf" dev="vda1" ino=26647789 scontext=system_u:system_r:foreman_proxy_t:s0 tcontext=system_u:object_r:puppet_etc_t:s0 tclass=file
type=AVC msg=audit(1415792587.199:1065): avc:  denied  { read } for  pid=23344 comm="ruby" name="puppet.conf" dev="vda1" ino=26647789 scontext=system_u:system_r:foreman_proxy_t:s0 tcontext=system_u:object_r:puppet_etc_t:s0 tclass=file
----
time->Wed Nov 12 11:43:07 2014
type=SYSCALL msg=audit(1415792587.199:1066): arch=c000003e syscall=16 success=no exit=-25 a0=c a1=5401 a2=7f0f233cb4e0 a3=1 items=0 ppid=1 pid=23344 auid=4294967295 uid=997 gid=995 euid=997 suid=997 fsuid=997 egid=995 sgid=995 fsgid=995 tty=(none) ses=4294967295 comm="ruby" exe="/usr/bin/ruby" subj=system_u:system_r:foreman_proxy_t:s0 key=(null)
type=AVC msg=audit(1415792587.199:1066): avc:  denied  { ioctl } for  pid=23344 comm="ruby" path="/etc/puppet/puppet.conf" dev="vda1" ino=26647789 scontext=system_u:system_r:foreman_proxy_t:s0 tcontext=system_u:object_r:puppet_etc_t:s0 tclass=file
----
time->Wed Nov 12 11:43:07 2014
type=SYSCALL msg=audit(1415792587.275:1067): arch=c000003e syscall=42 success=yes exit=0 a0=c a1=7f0f1409ddd0 a2=10 a3=7f0f233cb0f0 items=0 ppid=1 pid=23344 auid=4294967295 uid=997 gid=995 euid=997 suid=997 fsuid=997 egid=995 sgid=995 fsgid=995 tty=(none) ses=4294967295 comm="ruby" exe="/usr/bin/ruby" subj=system_u:system_r:foreman_proxy_t:s0 key=(null)
type=AVC msg=audit(1415792587.275:1067): avc:  denied  { name_connect } for  pid=23344 comm="ruby" dest=8140 scontext=system_u:system_r:foreman_proxy_t:s0 tcontext=system_u:object_r:puppet_port_t:s0 tclass=tcp_socket
----
time->Wed Nov 12 11:43:07 2014
type=SYSCALL msg=audit(1415792587.727:1068): arch=c000003e syscall=257 success=yes exit=12 a0=ffffffffffffff9c a1=7f0f14441fa0 a2=90800 a3=0 items=0 ppid=1 pid=23351 auid=4294967295 uid=997 gid=995 euid=997 suid=997 fsuid=997 egid=995 sgid=995 fsgid=995 tty=(none) ses=4294967295 comm="ruby" exe="/usr/bin/ruby" subj=system_u:system_r:foreman_proxy_t:s0 key=(null)
type=AVC msg=audit(1415792587.727:1068): avc:  denied  { open } for  pid=23351 comm="ruby" path="/etc/puppet/environments/example_env/modules" dev="vda1" ino=9019580 scontext=system_u:system_r:foreman_proxy_t:s0 tcontext=system_u:object_r:puppet_etc_t:s0 tclass=dir
type=AVC msg=audit(1415792587.727:1068): avc:  denied  { read } for  pid=23351 comm="ruby" name="modules" dev="vda1" ino=9019580 scontext=system_u:system_r:foreman_proxy_t:s0 tcontext=system_u:object_r:puppet_etc_t:s0 tclass=dir
----
time->Wed Nov 12 11:43:08 2014
type=SYSCALL msg=audit(1415792588.456:1069): arch=c000003e syscall=257 success=yes exit=13 a0=ffffffffffffff9c a1=7f0f1403b7f0 a2=90800 a3=0 items=0 ppid=1 pid=23357 auid=4294967295 uid=997 gid=995 euid=997 suid=997 fsuid=997 egid=995 sgid=995 fsgid=995 tty=(none) ses=4294967295 comm="ruby" exe="/usr/bin/ruby" subj=system_u:system_r:foreman_proxy_t:s0 key=(null)
type=AVC msg=audit(1415792588.456:1069): avc:  denied  { read } for  pid=23357 comm="ruby" name="manifests" dev="vda1" ino=26650307 scontext=system_u:system_r:foreman_proxy_t:s0 tcontext=unconfined_u:object_r:puppet_var_lib_t:s0 tclass=dir
</pre>

I remain unconvinced that this has any security benefits ;)

lzap commented:
> I remain unconvinced that this has any security benefits ;)

Sorry what you mean by _this_?

## 

Later,
 Lukas #lzap Zapletal

lzap commented:
> Seems to be missing support for the Puppet endpoint for class imports.

Pushed all the changes above.

Working on the imports.

## 

Later,
 Lukas #lzap Zapletal

lzap commented:
> Seems to be missing support for the Puppet endpoint for class imports.

Can't reproduce, when do you experience this?

## 

Later,
 Lukas #lzap Zapletal

lzap commented:
> Seems to be missing support for the Puppet endpoint for class imports.

Have you relabeled fcontexts?

```
bash -x policy-checkout/foreman-proxy-selinux-relabel
```

## 

Later,
 Lukas #lzap Zapletal

lzap commented:
Reproduced, fixed. There is now another boolean available. Are we all set? Are we there?!? ;-)

lzap commented:
@domcleal hey I know you're busy, I'd like to get this merged because of upcoming downstream bugzillas. I plan to work on many sat6 bugs in the incoming weeks. if you can't make it, we need someone else I guess. thanks!

domcleal commented on foreman-proxy.te:
```diff
@@ -0,0 +1,276 @@
+allow foreman_proxy_t self:udp_socket { getattr ioctl create connect };
+logging_send_syslog_msg(foreman_proxy_t)
+# echo service
+corenet_tcp_connect_inetd_child_port(foreman_proxy_t)
```
should this be around line 245 instead?

lzap replied:
On Wed, Nov 26, 2014 at 04:48:36AM -0800, Dominic Cleal wrote:

> should this be around line 245 instead?

Good catch, this explains lot of things :-)

  # ping support via ECHO TCP service
  ifdef(`distro_rhel6',`
    corenet_tcp_connect_inetd_child_port(foreman_proxy_t)
  ',`
    corenet_tcp_connect_echo_port(foreman_proxy_t)
  ')

The echo service port was introduced in RHEL7.

> ## 
> 
> Later,
>  Lukas #lzap Zapletal

domcleal commented on foreman-proxy.te:
```diff
@@ -0,0 +1,276 @@
+    sysnet_rw_dhcp_config(foreman_proxy_t)
+
+    # <dhcpd_admin>
+	gen_require(`
```
style nitpick, one less level of indentation to match the others above

lzap replied:
> style nitpick, one less level of indentation to match the others above

Oh `:retab` fixed it. Nasty tab chars.

## 

Later,
 Lukas #lzap Zapletal

domcleal commented on foreman-proxy.te:
```diff
@@ -0,0 +1,276 @@
+        sudo_exec(foreman_proxy_t)
+        allow foreman_proxy_t self:process { setsched setrlimit };
+        allow foreman_proxy_t self:capability { sys_resource setuid setgid };
+        logging_send_audit_msgs(foreman_proxy_t)
```
any ideas why these things are needed, when sudo has its own domain?

lzap replied:
> any ideas why these things are needed, when sudo has its own domain?

This does not do any transition, these are rules to actually allow our
domain to execute sudo, nothing else. We can't really change to sudo_t
it's meant for other purposes.

Basically these rules are necessary to do the change to `root`. I am not
going to add TYPE/ROLE support via `sudoers` file for this cut. We only
need few rules to operate properly, therefore transitions into domains
of the underlaying services would be an overkill.

## 

Later,
 Lukas #lzap Zapletal

domcleal replied:
Oh!  I thought it would transition.

Edit: oh yep, no transition rule.  Oh well.

domcleal commented:
I tried to build it, but it failed (this is an EL7 mock root).  It'd be great if you could provide scratch builds :)

<pre>
+ make NAME=targeted -f /usr/share/selinux/devel/Makefile DISTRO=rhel7
Compiling targeted foreman module
foreman.te:375: Warning: miscfiles_read_certs() has been deprecated, please use miscfiles_read_generic_certs() instead.
/usr/bin/checkmodule:  loading policy configuration from tmp/foreman.tmp
/usr/bin/checkmodule:  policy configuration loaded
/usr/bin/checkmodule:  writing binary representation (version 17) to tmp/foreman.mod
Creating targeted foreman.pp policy package
Compiling targeted foreman-proxy module
foreman-proxy.te:182: Warning: miscfiles_read_certs() has been deprecated, please use miscfiles_read_generic_certs() instead.
foreman-proxy.te":19:WARNING 'unrecognized character' at token '@' on line 1002:
                module foreman-proxy @@VERSION@@;
#line 19
foreman-proxy.te":19:WARNING 'unrecognized character' at token '@' on line 1002:
                module foreman-proxy @@VERSION@@;
#line 19
foreman-proxy.te":19:ERROR 'syntax error' at token 'VERSION' on line 1002:
                module foreman-proxy @@VERSION@@;
#line 19
/usr/bin/checkmodule:  error(s) encountered while parsing configuration
/usr/bin/checkmodule:  loading policy configuration from tmp/foreman-proxy.tmp
rm tmp/foreman.mod.fc tmp/foreman.mod
RPM build errors:
make: *** [tmp/foreman-proxy.mod] Error 1
error: Bad exit status from /var/tmp/rpm-tmp.8VdxWk (%build)
    Bad exit status from /var/tmp/rpm-tmp.8VdxWk (%build)
</pre>

domcleal commented:
Looking a hell of a lot better, thanks.

By "this" I meant trying to confine the proxy.. I think it might only help in certain circumstances, like a TFTP-only proxy.  With these huge polices behind some of the booleans, I'm not sure it confines all that much in a more common all-in-one setup.

lzap commented:
Please finish with the review and I will send the builds of course.

To build on a remote system, do:

  make remote-load DISTRO=rhel6 HOST=fsix.zzz.lan

Which does something like:

  sed -i s/@@VERSION@@/99.999/ *.te && make -f /usr/share/selinux/devel/Makefile load DISTRO=rhel7

## 

Later,
 Lukas #lzap Zapletal

domcleal commented on foreman-proxy.te:
```diff
@@ -0,0 +1,276 @@
+    ')
+    # connect to puppet
+    corenet_tcp_connect_puppet_port(foreman_proxy_t)
+    # read configs and manage puppet files
```
These need to be outside of this tunable, they're used in non-Puppet API mode too, if you don't have directory environments.  If you have foreman_proxy_manage_puppetca enabled then it works as that also grants similar access, but they can be used separately.

domcleal replied:
Probably best to have a new tunable to represent the "puppet" module of the smart proxy ("foreman_proxy_manage_puppet"), and replace this use_puppet_api tunable as it's the modern/future default behaviour.

If you want to keep the use_puppet_api tunable for older (pre Puppet 3.6) versions then I guess you could, but it seems an unnecessary extra.

lzap replied:
So do I understand you right just to rename `foreman_proxy_use_puppet_api` to `foreman_proxy_manage_puppet` ?

domcleal replied:
Yes :)

(Sorry, you received my entire train of thought...)

domcleal commented:
Oh, please ignore the comment about build failures.  I hadn't used your packaging branch, duh.

domcleal commented:
Okay, looking good.  Could you push the last updates we talked about?  (The ICMP bit, indentation)

Anything else on your mind, or is it OK to merge?

lzap commented:
Ok all is fixed. Pushed.

You ask me if we are good to go? For God sake, I beg you on my knees - please do merge :-)

domcleal commented:
Merged as f4a857f for Foreman 1.8.  Thanks @lzap!
